numba/targets/arrayobj.py

```diff
@@ -4857,21 +4859,189 @@ def np_sort_impl(a):
+
+        return tup
+
+    return function_sig, codegen
+
+
+@overload_method(types.Array, "argsort")
+@overload(np.argsort)
+def np_argsort(a, axis=-1, kind="quicksort"):
+    # Unpack TypeRef
+    axis = getattr(axis, "instance_type", axis)
+    kind = getattr(kind, "instance_type", kind)
+
+    # Unpack Omitted
+    axis = getattr(axis, "value", axis)
+    kind = getattr(kind, "value", kind)
+
+    # Wrap python types
+    if isinstance(axis, int):
+        axis = types.IntegerLiteral(axis)
+    if isinstance(kind, str):
+        kind = types.StringLiteral(kind)
+
+    if not (isinstance(a, types.Array) and
+            isinstance(kind, types.StringLiteral)):
+        return
+
+    arytype = a
+
     sort_func = get_sort_func(kind=kind.literal_value,
                               is_float=isinstance(arytype.dtype, types.Float),
                               is_argsort=True)
 
-    def array_argsort_impl(arr):
-        return sort_func(arr)
+    # If axis is None, return 1D argsort on flattened array.
+    if axis in (None, types.none):
+        def array_argsort_impl(a, axis=-1, kind="quicksort"):
+            return sort_func(a.flatten())
+        return array_argsort_impl
+
+    if not isinstance(axis, types.Integer):
+        return
+
+    if isinstance(axis, types.IntegerLiteral):
+        axis_value = axis.literal_value
+        if axis_value < 0:
+            axis_value += arytype.ndim
+
+        if axis_value >= arytype.ndim:
+            raise ValueError("axis {} is out of bounds".format(axis_value) +
+                             " for array of dimension {}".format(arytype.ndim))
+        axis_is_const = True
+    else:
+        axis_value = 0
+        axis_is_const = False
+
+    # Optimized implementation for 1D arrays
+    if arytype.ndim == 1:
+        def array_argsort_impl(a, axis=-1, kind="quicksort"):
+            axis_val = axis_value if axis_is_const else axis
+            assert axis_val in (-1, 0)
+            return sort_func(a)
+    else:
+        def gen_fill_loop(axis_val):
+            @register_jitable
+            def impl(arr):
+                max_idx = 1
+                for i, v in enumerate(arr.shape):
+                    if i != axis_val:
+                        max_idx *= v
+
+                result = np.empty(arr.shape, types.intp)
+                for idx in range(max_idx):
+                    slice_tuple = _gen_slice_tuple(arr.shape, idx, axis_val)
+                    result[slice_tuple] = sort_func(arr[slice_tuple])
+                return result
+            return impl
+
+        fill_loop_const = gen_fill_loop(axis_value)
+        fill_loop_0 = gen_fill_loop(0)
+        fill_loop_1 = gen_fill_loop(1)
+        fill_loop_2 = gen_fill_loop(2)
+        fill_loop_3 = gen_fill_loop(3)
+        fill_loop_n1 = gen_fill_loop(-1)
+
+        def array_argsort_impl(a, axis=-1, kind="quicksort"):
+            if axis_is_const:
+                return fill_loop_const(a)
```

Comment:
The overhead of compiling this is quite high because the definition is requiring a fresh `fill_loop_X` that is being compiled for each signature of `argsort`.  It would suggest splitting the implementation so that there's one version for `axis_is_const` being `True` and one for otherwise. It'd avoid compiling a new version for each constant.